neo/Network/P2P/ProtocolHandler.cs

```diff
@@ -21,9 +22,32 @@ public class SetVersion { public VersionPayload Version; }
         public class SetVerack { }
         public class SetFilter { public BloomFilter Filter; }
 
+        private class LimitedSet<T>
+        {
+            private int maxCapacity;
+            private OrderedDictionary orderedDictionary;
+
+            public LimitedSet(int maxCapacity = 100)
+            {
+                if (maxCapacity <= 0) throw new ArgumentOutOfRangeException(nameof(maxCapacity));
+
+                this.maxCapacity = maxCapacity;
+                this.orderedDictionary = new OrderedDictionary(maxCapacity);
+            }
+
+            public bool Add(T item)
+            {
+                if (orderedDictionary.Contains(item)) return false;
+                if (orderedDictionary.Count >= maxCapacity)
+                    orderedDictionary.RemoveAt(0);
```

Comment:
FIFO, right?

Reply:
Yes, but the existing FIFO cache (https://github.com/neo-project/neo/blob/master/neo/IO/Caching/FIFOCache.cs) is key-value based.

Reply:
I think @vncoelho was just asking to clarify that your implementation here is FIFO. And it is so, I think we can resolve this comment.

I agree that you shouldn’t use the existing implementation of the `FIFOCache` class since it has locking that you don’t need here, plus what it does for removing when at capacity is not optimal as it sorts each time to remove the oldest (and it doesn’t use a data structure that keeps the items sorted). Its approach wouldn’t be so bad if it used batch removal. We should actually improve the `Cache` class implementation also for how it handles the at capacity situation.